### PR TITLE
feat: datasource access to allow more granular access to tables on SQL Lab

### PR DESCRIPTION
**Pull request by victorarbuesmallada**
### SUMMARY

We'd like to have a more granular access to tables on SQL Lab, which currently can only be done at schema or database level. This PR would then tackle the [story](https://github.com/apache/superset/issues/18014) we raised a couple of days ago.

### TESTING INSTRUCTIONS
Assign a datasource permission to a role and check that users with that role can only query that particular datasource.

### ADDITIONAL INFORMATION
- [x] Has associated issue: https://github.com/apache/superset/issues/18014
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by geido** [comment]
Hey @Painyjames it's great that you went forward and implemented this. Let me ping a few people who can review this PR!

**Comment by betodealmeida** [comment]
I created a test user `sqllab` and assigned it to the role `sql_lab`, and assigned the permission `datasource access on [examples].[messages](id:10)` to that role.

But I wasn't able to query the table:

```sql
SELECT * FROM examples.messages
```

On the browser console I can see the error:

```json
{
  "errors": [
    {
      "message": "You need access to the following tables: `examples.messages`,\n            `all_database_access` or `all_datasource_access` permission",
      "error_type": "TABLE_SECURITY_ACCESS_ERROR",
      "extra": {
        "link": "",
        "tables": [
          "examples.messages"
        ]
      }
    }
  ]
}
```

Maybe @dpgaspar can shed some light here?

**Comment by Painyjames** [comment]
> I created a test user `sqllab` in the ephemeral environment above (password `sqllab`) and assigned it to the role `sql_lab`, and assigned the permission `datasource access on [examples].[messages](id:10)` to that role.
> 
> But I wasn't able to query the table:
> 
> ```sql
> SELECT * FROM examples.messages
> ```

I think the reason why this might not works is because the permission is `datasource access on [database].[table]` but on the sql query we have `SELECT * FROM schema.table`.

In the acceptance tests we have against our patched superset container, we have the following:
- permission `datasource access on [postgres].[audit_logs]`
- sql editor query `SELECT * FROM public.audit_logs`

This works fine cause we only have one schema on our database, but we wouldn't know what would happen if we had several schemas with the same table on the same database.

In any case, this is still better than having only the schema permission that would grant access to all the tables under that schema, but honestly maybe we should consider generating permissions which are like this:
- `datasource access on [database].[schema].[table]`

**Comment by Painyjames** [comment]
I've just test it out on the [test environment ](http://54.200.220.32:8080/)and, even though couldn't load the schema, I was able to run a SQL statement against the allowed table.
Screenshot 2022-01-20 at 09 16 25

**Comment by Painyjames** [comment]
Adding as well what we have locally with this PR's changes
Screenshot 2022-01-20 at 09 44 57

File: tests/integration_tests/core_tests.py
Comment (dpgaspar) [inline tests/integration_tests/core_tests.py:191]: can you add a test where the table is not allowed also?

**Review by dpgaspar — COMMENTED** [review]
Quick pass, left a comment. This is a delicate change, we should make sure this is backward compatible

File: tests/integration_tests/core_tests.py
Comment (Painyjames) [reply tests/integration_tests/core_tests.py:191]: added 👍

**Comment by Painyjames** [comment]
@villebro @betodealmeida @zhaoyongjie are you guys happy with this PR? :)

File: superset/databases/filters.py
Comment (villebro) [inline superset/databases/filters.py:38]: nit: not something you introduced, but while we're touching up the code and for readability's sake, it would be nice to use a `NamedTuple` or `@dataclass` as the return type in `unpack_perm` so we can replace `[0]` with `.database_name` here and `[1]` with `.schema_name` when requesting the schema name.

**Review by villebro — COMMENTED** [review]
Really nice improvement! I left a minor readability comment, but the changes LGTM at face value. I agree with @dpgaspar that this is a delicate change, so we should make sure to test this properly. I will do some testing tomorrow.

**Comment by villebro** [comment]
@Painyjames I just spun up an ephemeral envirnonment for testing - if it's not too much trouble, do you think it would be possible to add a few test users with relevant perms (at leat one with and one without datasource access) and instructions for testing so we can validate the changes here?

**Comment by Painyjames** [comment]
I've created a couple of users:
- `testcovidvaccines` (same username and password), which has access to the `covid_vaccines` datasource.
Screenshot 2022-02-01 at 22 27 52
- `testnoaccess` (same username and password), which has no access to any datasource.

**Comment by Painyjames** [comment]
Worth mentioning that, with the datasource access user,  somehow I cannot select the schema and table on the dropdowns, although querying the allowed datasource is still possible.
This doesn't happen locally when the datasource is coming from a postgres database (instead of sql lite) as you can see [here](https://github.com/apache/superset/pull/18064#issuecomment-1017292048).

**Comment by villebro** [comment]
> Worth mentioning that, with the datasource access user, somehow I cannot select the schema and table on the dropdown lists, although querying the allowed datasource is still possible. This doesn't happen locally when the datasource is coming from a postgres database (instead of sql lite) as you can see [here](https://github.com/apache/superset/pull/18064#issuecomment-1017292048).

I believe this is an unrelated bug and might be resolved by #18564

**Comment by mayurnewase** [comment]
in the issue, the requirement was to allow querying on physical table on db but datasource permission here is used for datasource(virtual/physical) layer in superset.
So if datasource is not created here but table exists in db, is user allowed to query it?

**Comment by mayurnewase** [comment]
Also I think the reason it doesn't show schema in the dropdown for sqlite is when examples are loaded in sqlite, this attribute `https://docs.sqlalchemy.org/en/14/core/reflection.html#sqlalchemy.engine.reflection.Inspector.default_schema_name` gives `None` instead of `main` so those schemas get filtered when getting schemas accessible by user.

**Review by villebro — APPROVED** [review]
Tested and I agree we can leave the sqlite issue to be fixed in a follow-up PR. As the added functionality is well in line with the current logic, I feel this is good to go. @mayurnewase do you feel there's risk in merging this? I would also love to get an approval from @dpgaspar and/or @betodealmeida before merging.

**Comment by mayurnewase** [comment]
I tested on postgres and worked as intended.

File: superset/databases/filters.py
Comment (zhaoyongjie) [inline superset/databases/filters.py:49]:
```suggestion
               self.model.database_name.in_([*schema_access_databases, *datasource_access_databases]),
```

**Review by zhaoyongjie — APPROVED** [review]
no blocking suggestion, other LGTM.

**Comment by zhaoyongjie** [comment]
@Painyjames Could you fix this code style issue? thanks!

File: superset/databases/filters.py
Comment (dpgaspar) [inline superset/databases/filters.py:29]: nit: any reason why this can't be `@staticmethod` ?

**Pull request merged**
